backend/api/Areas/Project/Controllers/SnapshotController.cs

```diff
@@ -0,0 +1,151 @@
+            _mapper = mapper;
+        }
+        #endregion
+
+        #region Endpoints
+        /// <summary>
+        /// Get the project snapshot for the specified 'id'.
+        /// </summary>
+        /// <param name="id"></param>
+        /// <returns></returns>
+        [HttpGet("{id:int}")]
+        [HasPermission(Permissions.AdminProjects)]
+        [Produces("application/json")]
+        [ProducesResponseType(typeof(ProjectSnapshotModel), 200)]
+        [ProducesResponseType(typeof(ErrorResponseModel), 400)]
+        [SwaggerOperation(Tags = new[] { "project" })]
+        public IActionResult GetProjectSnapshot(int id)
+        {
+            var snapshot = _pimsService.ProjectSnapshot.Get(id);
+            return new JsonResult(_mapper.Map<ProjectSnapshotModel>(snapshot));
+        }
+
+        /// <summary>
+        /// Add the project snapshot to the datasource.
+        /// </summary>
+        /// <param name="model"></param>
+        /// <returns></returns>
+        [HttpPost]
+        [HasPermission(Permissions.AdminProjects)]
+        [Produces("application/json")]
+        [ProducesResponseType(typeof(ProjectSnapshotModel), 201)]
+        [ProducesResponseType(typeof(ErrorResponseModel), 400)]
+        [SwaggerOperation(Tags = new[] { "project" })]
+        public IActionResult AddProjectSnapshot(ProjectSnapshotModel model)
+        {
+            var snapshot = _pimsService.ProjectSnapshot.Add(_mapper.Map<Entity.ProjectSnapshot>(model));
+            return new JsonResult(_mapper.Map<ProjectSnapshotModel>(snapshot));
+        }
+
+        /// <summary>
+        /// Update the project snapshot metadata in the datasource.
+        /// </summary>
+        /// <param name="model"></param>
+        /// /// <param name="id"></param>
+        /// <returns></returns>
+        [HttpPut("{id}")]
+        [HasPermission(Permissions.AdminProjects)]
+        [Produces("application/json")]
+        [ProducesResponseType(typeof(ProjectSnapshotModel), 201)]
+        [ProducesResponseType(typeof(ErrorResponseModel), 400)]
+        [SwaggerOperation(Tags = new[] { "project" })]
+        public IActionResult UpdateProjectSnapshot(int id, ProjectSnapshotModel model)
+        {
+            var snapshot = _pimsService.ProjectSnapshot.Update(_mapper.Map<Entity.ProjectSnapshot>(model));
+            return new JsonResult(_mapper.Map<ProjectSnapshotModel>(snapshot));
+        }
+
+        /// <summary>
+        /// Refresh the project snapshot in the datasource.
+        /// </summary>
+        /// <param name="id"></param>
+        /// <returns></returns>
+        [HttpGet("refresh/{id}")]
+        [HasPermission(Permissions.AdminProjects)]
+        [Produces("application/json")]
+        [ProducesResponseType(typeof(ProjectSnapshotModel), 201)]
+        [ProducesResponseType(typeof(ErrorResponseModel), 400)]
+        [SwaggerOperation(Tags = new[] { "project" })]
+        public IActionResult RefreshProjectSnapshot(int id)
```

Comment:
This does a 'soft' refresh of the snapshot pointed to by this id. The user needs to save the refreshed data if they want it to persist.

Reply:
I'm wondering if it makes sense to call this a refresh.  It's a GET, so non-permanent.  Wouldn't it just make sense that this is essentially fetching the latest information for the specified "from snapshot id"?

Reply:
I've noted below perhaps this makes more sense to be called **Generate**.